kafka-rest/src/main/java/io/confluent/kafkarest/resources/v3/ProducerMetrics.java

```diff
@@ -50,339 +40,226 @@
+    this.requestSizeSensorName = sensorNamePrefix + REQUEST_SIZE_SENSOR_NAME;
+    this.responseSensorName = sensorNamePrefix + RESPONSE_SENSOR_NAME;
+    setupSensors();
+  }
 
-    void recordRequest() {
-      recordMetric(fullyQualifiedRequestSensor, 1.0);
-    }
+  private static Metrics createMetrics(RestMetricsContext metricsContext, Time time) {
+    JmxReporter reporter = new JmxReporter();
+    reporter.contextChange(metricsContext);
+    return new Metrics(
+        new MetricConfig()
+            .samples(NUM_SAMPLES)
+            .timeWindow(SAMPLE_WINDOWS_MS, TimeUnit.MILLISECONDS)
+            .recordLevel(RECORDING_LEVEL),
+        singletonList(reporter),
+        time,
+        metricsContext);
+  }
 
-    void recordRequestSize(double value) {
-      recordMetric(fullyQualifiedRequestSizeSensor, value);
-    }
+  private void setupSensors() {
+    // Clear out any previous metrics and sensors that might have been present
+    removeExistingMetrics();
 
-    private void recordMetric(String sensorName, double value) {
-      Sensor sensor = metrics.getSensor(sensorName);
-      if (sensor != null) {
-        sensor.record(value);
-      }
-    }
+    // request metrics
+    setupRequestSensor();
+    setupRequestSizeSensor();
 
-    /**
-     * Create the name of a metric that belongs to this group and has the group's tags.
-     *
-     * @param template the name template for the metric; may not be null
-     * @return the metric name; never null
-     * @throws IllegalArgumentException if the name is not valid
-     */
-    private MetricName metricName(MetricNameTemplate template) {
-      return metrics.metricInstance(template, beanCoordinate.tags);
-    }
+    // response metrics
+    setupResponseSensor();
+    setupRecordErrorSensor();
+    setupRequestLatencySensor();
 
-    /**
-     * Get or create a sensor with the given unique name and no parent sensors. This uses a default
-     * recording level of INFO.
-     *
-     * @param name The sensor name
-     * @return The sensor
-     */
-    private synchronized Sensor sensor(String name) {
-      Sensor sensor = metrics.sensor(name);
-      return sensor;
-    }
+    log.info("Successfully registered kafka-rest produce metrics with JMX");
+  }
 
-    /** Remove all sensors and metrics associated with this group. */
-    @Override
-    public synchronized void close() {
-      for (MetricName metricName : new HashSet<>(metrics.metrics().keySet())) {
-        if (metricName.group().equals(beanCoordinate.beanName)
-            && beanCoordinate.tags.equals(metricName.tags())) {
-          metrics.removeMetric(metricName);
-        }
+  private void removeExistingMetrics() {
+    for (MetricName metricName : metrics.metrics().keySet()) {
+      if (metricName.group().equals(GROUP_NAME) && METRIC_TAGS.equals(metricName.tags())) {
+        metrics.removeMetric(metricName);
       }
     }
   }
```

Comment:
When are we expecting this not to match any metric? And additionally, if this is only used when setting up sensors, and we are using these solely for aggregate metrics (not e.g. per-tenant metrics), and we are treating the `ProducerMetrics` instance as a commonly provided singleton, do we really need that step?

Reply:
Done. Agreed. Seems futile, specially given this will be a singleton.